File: swordTo_offer/42.cpp
#include "header.h"
//连续子数组的最大和 DP
class Solution1 {
public:
    int maxSubArray(vector<int>& nums) {
        int n = nums.size();
        vector<int> dp(n + 1, 0);
        int ans = nums[0];
        for (int i = 1; i <= n; ++i) {
            dp[i] = max(dp[i - 1] + nums[i - 1], nums[i - 1]);
            ans = max(dp[i], ans);
        }
        return ans;
    }
};

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int n = nums.size();
        int pre = 0;
        int ans = nums[0];
        for (int i = 1; i <= n; ++i) {
            pre = max(pre + nums[i - 1], nums[i - 1]);
            ans = max(pre, ans);
        }
        return ans;
    }
};

File: swordTo_offer/42.py
from typing import List
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [0] * (n + 1)
        ans = nums[0]
        for i in range(1, n + 1) :
            dp[i] = max(dp[i - 1] + nums[i - 1], nums[i - 1])
            ans = max(dp[i], ans)
        return ans


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        n = len(nums)
        pre = 0
        ans = nums[0]
        for i in range(1, n + 1) :
            pre = max(pre + nums[i - 1], nums[i - 1]);
            ans = max(pre, ans)
        return ans